src/lib.rs

```diff
@@ -3069,6 +3071,122 @@ pub fn digraph_core_number(
     _core_number(py, &graph.graph)
 }
 
+/// Find the edges in the minimum spanning tree or forest of a graph
+/// using Kruskal's algorithm.
+///
+/// :param PyGraph graph: Undirected graph
+/// :param weight_fn: A callable object (function, lambda, etc) which
+///     will be passed the edge object and expected to return a ``float``. This
+///     tells retworkx/rust how to extract a numerical weight as a ``float``
+///     for edge object. Some simple examples are::
+///
+///         minimum_spanning_edges(graph, weight_fn: lambda x: 1)
+///
+///     to return a weight of 1 for all edges. Also::
+///
+///         minimum_spanning_edges(graph, weight_fn: lambda x: float(x))
+///
+///     to cast the edge object as a float as the weight.
+/// :param float default_weight: If ``weight_fn`` isn't specified this optional
+///     float value will be used for the weight/cost of each edge.
+///
+/// :returns: The :math:`N - |c|` edges of the Minimum Spanning Tree (or Forest, if :math:`|c| > 1`)
+///     where :math:`N` is the number of nodes and :math:`|c|` is the number of connected components of the graph
+/// :rtype: WeightedEdgeList
+#[pyfunction(weight_fn = "None", default_weight = "1.0")]
+#[text_signature = "(graph, weight_fn=None, default_weight=1.0)"]
+pub fn minimum_spanning_edges(
+    py: Python,
+    graph: &graph::PyGraph,
+    weight_fn: Option<PyObject>,
+    default_weight: f64,
+) -> PyResult<WeightedEdgeList> {
+    let mut subgraphs = UnionFind::<usize>::new(graph.graph.node_bound());
+
+    let mut edge_list: Vec<(f64, EdgeReference<PyObject>)> = Vec::new();
+    for edge in graph.edge_references() {
+        edge_list.push((
+            weight_callable(py, &weight_fn, &edge.weight(), default_weight)?,
+            edge,
+        ));
+    }
+
+    edge_list.sort_unstable_by(|a, b| {
+        let weight_a = a.0;
+        let weight_b = b.0;
+        if weight_a == weight_b {
+            Ordering::Equal
+        } else if weight_a < weight_b {
+            Ordering::Less
+        } else if weight_a > weight_b {
+            Ordering::Greater
+        } else if weight_a.is_nan() && weight_b.is_nan() {
+            Ordering::Equal
+        } else if weight_a.is_nan() {
+            Ordering::Greater
+        } else {
+            Ordering::Less
+        }
```

Comment:
If we're willing to say `NaN` isn't allowed I think we can just use `a.partial_cmp(b).unwrap());` here. This is actually what the rust `sort_by` docs point out. I think this is a fair compromise and we probably can just return a python exception in the edge references loop above if `weight_callable` returns a nan (we also should have a test for that).

Reply:
Done (with test added).

Also, I opted to use `a.partial_cmp(b).unwrap_or(Orderding::Less)` because it provides a slight performance boost (Rust inlines code better if there's no `panic!()`).